yt/visualization/accumulators.py

```diff
@@ -0,0 +1,348 @@
+import numpy as np
+
+import unyt
+from unyt.array import ustack
+
+from yt import YTArray
+from yt.funcs import ensure_list
+from yt.utilities.amr_kdtree.api import AMRKDTree
+
+
+def _accumulate_vector_field(path, field_vals):
+    r"""
+    This function integrates the given vector field along the given
+    path p. The integral is done in a piecewise manner
+    (segment-by-segment) so as to be able to store the accumulated
+    values from each of the previous segments.
+
+    For a vector field, the line integral is:
+
+    ..math::
+
+        I = \int_C \vec{a} \cdot d\vec{r}
+
+    where :math:`C` is the path being integrated along, :math:`\vec{a}` is
+    the vector field being integrated, and :math:`d\vec{r}` points along the
+    path :math:`C`. This is equivalent to:
+    
+    ..math::
+
+        \lim_{N \rightarrow \infty} \sum_{p=1}^N \vec{a}(x_p, y_p, ...) \cdot \Delta \vec{r}_p
+        
+    with the understanding that :math:`\Delta \vec{r}_p \rightarrow 0` as
+    :math:`N \rightarrow \infty`.
+        
+    The vector pointing along the segment connecting two adjacent points is:
+    
+    ..math::
+    
+        \Delta\vec{r}_p = (\vec{r}_{p+1} - \vec{r}_p)
+        
+    so the full dot product can be written out as:
+
+    ..math::
+        I = \lim_{N \rightarrow \infty}\sum_{p=1}^N\sum_{i=1}^n a_{p,i}(r_{p+1,i}-r_{p,i})
+        
+    where :math:`n` is the number of dimensions.
+
+    This can be done as a matrix operation. If :math:`\vec{r} \equiv \vec{r}_{p+1} - \vec{r}_p`
+    then the above sum is (leaving out the limit):
+    
+    ..math::
+    
+        I = \vec{a}_1 \dot \vec{r}_1 + \vec{a}_2 \cdot \vec{r}_2 \ldots \vec{a}_N \cdot \vec{r}_N
+        
+    If we write the matrix :math:`A = ([a1x, a1y, ...], [a2x, a2y, ...], ...)`
+    and :math:`R = ([r1x, r1y, ...], [r2x, r2y, ...], ...)^T`, then the dot
+    products in the sum are the diagonal elements of the resulting matrix
+    multiplication :math:`AR`. The accumulation is then obtained by doing a
+    cumsum of this diagonal.
+
+    Parameters
+    ----------
+    p : YTArray 
+        The path to be integrated along
+
+    field_vals : YTArray 
+        An array containing the components of the vector field to be
+        integrated. The values are sampled at the starting point of
+        each path segment as well as the endpoint for the last segment
+
+    Returns
+    -------
+        accum : YTArray 
+            The cumulative value of the field integral at each path
+            segment
+    """
+    # https://en.wikipedia.org/wiki/Line_integral
+    # np.dot doesn't combine units correctly and there does not appear to be a
+    # unyt implementation of dot, so units will be handled manually
+    # for the time being
+    accum = np.cumsum(np.diag(np.dot(field_vals[:-1].d, (path[1:].d - path[:-1].d).T)))
```

Comment:
`path` is unitless, right?  That's why this works?

Reply:
Yeah, path is unitless. Looking at it now, this probably isn't a good way of doing it. I assume that the path is in code units. I think I need to handle the units better, as well as make path a unyt array. I'll do that.

Reply:
No, I was wrong. path either has units specified by the user or, if no units are given by the user, then code length units are assumed. The above construction was done because np.dot was removing the units completely and I did not find a unyt implementation of np.dot. As such, I used .d to just access the quantities in the path and field arrays and then manually handled the units. There might be a better way to do that, though.

Reply:
The units are manually added in on line 82